torch_xla/experimental/pallas_kernels/multi_queries_paged_attention_kernel.py

```diff
@@ -0,0 +1,612 @@
+      pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.ANY),
+      None,
+      pl.BlockSpec(memory_space=pltpu.TPUMemorySpace.ANY),
+      None,
+  ]
+
+  # out_spec
+  o_specs = pl.BlockSpec(  # Should be the same as q_block_spec
+      # q_ref.shape=[1, num_q_heads_per_kv_head, num_queries_per_compute_block, head_dim]
+      (1, num_q_heads_per_kv_head, num_queries_per_compute_block, head_dim),
+      qo_index_map,  # map from grid idx to q's starting index
+  )
+
+  # lm_index_map is same as qo_index_map
+  def lm_index_map(batch_index, kv_head_index, q_seq_blk_idx, *_):
+    return (batch_index, kv_head_index, q_seq_blk_idx, 0)
+
+  out_specs = [
+      o_specs,
+      pl.BlockSpec((1, num_q_heads_per_kv_head, num_queries_per_compute_block,
+                    MIN_BLOCK_SIZE), lm_index_map),  # l
+      pl.BlockSpec((1, num_q_heads_per_kv_head, num_queries_per_compute_block,
+                    MIN_BLOCK_SIZE), lm_index_map),  # m
+  ]
+
+  # scratch space. Note k_pages.shape=[num_kv_heads, total_num_pages, page_size, head_dim]
+  l_scratch = pltpu.VMEM(
+      (num_q_heads_per_kv_head, num_queries_per_compute_block, MIN_BLOCK_SIZE),
+      jnp.float32)
+  m_scratch = pltpu.VMEM(
+      (num_q_heads_per_kv_head, num_queries_per_compute_block, MIN_BLOCK_SIZE),
+      jnp.float32)
+  acc_scratch = pltpu.VMEM(
+      (num_q_heads_per_kv_head, num_queries_per_compute_block, head_dim),
+      jnp.float32)
+  scratch_shapes = [
+      pltpu.VMEM(
+          (
+              2,  # For double buffering during DMA copies.
+              num_kv_pages_per_compute_block,
+              page_size,
+              head_dim,
+          ),
+          k_pages.dtype,
+      ),  # k_pages buffer, k_pages.shape=[num_kv_heads, total_num_pages, page_size, head_dim]
+      None,  # k_scales_pages=None
+      pltpu.VMEM(
+          (
+              2,  # For double buffering during DMA copies.
+              num_kv_pages_per_compute_block,
+              page_size,
+              head_dim,
+          ),
+          v_pages.dtype,
+      ),  # v_pages buffer
+      None,  # v_scales_pages=None
+      pltpu.SemaphoreType.DMA,
+      l_scratch,
+      m_scratch,
+      acc_scratch,
+  ]
+
+  kernel = pl.pallas_call(
+      functools.partial(
+          paged_flash_attention_kernel,
+          pages_per_sequence=pages_per_sequence,
+          batch_size=batch_size,
+          num_kv_pages_per_compute_block=num_kv_pages_per_compute_block,
+          num_queries_per_compute_block=num_queries_per_compute_block,
+          mask_value=mask_value,
+          query_len=query_len),
+      grid_spec=pltpu.PrefetchScalarGridSpec(
+          num_scalar_prefetch=4,
+          in_specs=in_specs,
+          out_specs=out_specs,
+          grid=grid,
+          scratch_shapes=scratch_shapes,
+      ),
+      compiler_params=pltpu.TPUCompilerParams(
+          dimension_semantics=(
+              "parallel",
+              "parallel",
+              "parallel",
```

Comment:
As the kv blk prefetching for next iteration is done inside the kernel according to the `compute_block_indices`, these grid dims cannot be `"parallel"`. That's why the original implementation has to manually split out a `"parallel"` grid for megacore.

Reply:
Good call! Thanks.